src/number_theory/basic.lean

```diff
@@ -41,4 +50,247 @@ begin
+  geom_sum₂_eq_zero hp hxy₁ hxy₂ hn, zero_add]
+
+variables {p : ℕ} (hp : prime (p : R)) (hp1 : p % 2 = 1)
+  {x y : R} (hxy₁ : ↑p ∣ x - y) (hxy₂ : ¬↑p ∣ x * y)
+include hp hp1 hxy₁ hxy₂
+
+lemma geom_sum₂_eq_one : multiplicity ↑p (geom_sum₂ x y p) = 1 :=
+begin
+  apply multiplicity.eq_some_iff.mpr,
+  split,
+  { rw ← mem_span_singleton at *,
+    rw ← ideal.quotient.eq at hxy₁,
+    rw [pow_one, ← eq_zero_iff_mem, ring_hom.map_geom_sum₂, hxy₁, geom_sum₂_self],
+    apply mul_eq_zero_of_left,
+    rw [← ring_hom.map_nat_cast (mk _), eq_zero_iff_mem, mem_span_singleton] },
+  { let I : ideal R := span {p ^ 2},
+    have hp' : (p : I.quotient) ^ 2 = 0,
+    { rw [← ring_hom.map_nat_cast (mk I), ← ring_hom.map_pow,
+        eq_zero_iff_mem, mem_span_singleton] },
+    cases hxy₁ with k hk,
+    rw [← mem_span_singleton, ← eq_zero_iff_mem, ring_hom.map_geom_sum₂,
+      sub_eq_iff_eq_add'.mp hk, ring_hom.map_add, ring_hom.map_mul, ring_hom.map_nat_cast,
+      aux₂ hp' hp1, ← ring_hom.map_nat_cast (mk I), ← ring_hom.map_pow, ← ring_hom.map_mul,
+      eq_zero_iff_mem, mem_span_singleton, pow_two, mul_dvd_mul_iff_left hp.1],
+    intro h,
+    exact hxy₂ (dvd_mul_of_dvd_right (hp.dvd_of_dvd_pow h) _) }
+end
+
+lemma pow_prime_sub_pow_prime :
+  multiplicity ↑p (x ^ p - y ^ p) = multiplicity ↑p (x - y) + 1 :=
+by rw [← geom_sum₂_mul, multiplicity.mul hp,
+    geom_sum₂_eq_one hp hp1 hxy₁ hxy₂, add_comm]
+
+lemma pow_prime_pow_sub_pow_prime_pow (a : ℕ) :
+  multiplicity ↑p (x ^ p ^ a - y ^ p ^ a) = multiplicity ↑p (x - y) + a :=
+begin
+  induction a with a h_ind,
+  { rw [enat.coe_zero, add_zero, pow_zero, pow_one, pow_one] },
+  rw [← nat.add_one, enat.coe_add, enat.coe_one, ← add_assoc,
+    ← h_ind, pow_succ', pow_mul, pow_mul],
+  apply pow_prime_sub_pow_prime hp hp1,
+  { rw ← geom_sum₂_mul,
+    exact dvd_mul_of_dvd_right hxy₁ _ },
+  { rw ← mul_pow,
+    intro h,
+    exact hxy₂ (hp.dvd_of_dvd_pow h) }
+end
+
+end integral_domain
+
+section lifting_the_exponent
+
+variables {p : ℕ} (hp : p.prime) (hp1 : p % 2 = 1)
+include hp hp1
+
+/-- **Lifting the exponent lemma** for odd primes -/
+theorem int.pow_sub_pow {x y : ℤ} (hxy₁ : ↑p ∣ x - y) (hxy₂ : ¬↑p ∣ x * y) (n : ℕ) :
+  multiplicity ↑p (x ^ n - y ^ n) = multiplicity ↑p (x - y) + multiplicity p n :=
+begin
+  cases n,
+  { rw [multiplicity.zero, add_top, pow_zero, pow_zero, sub_self, multiplicity.zero] },
+  have h : (multiplicity _ _).dom := finite_nat_iff.mpr ⟨hp.ne_one, n.succ_pos⟩,
+  rcases eq_some_iff.mp (enat.coe_get h).symm with ⟨⟨k, hk⟩, hpn⟩,
+  conv_lhs { rw [hk, pow_mul, pow_mul] },
+  rw nat.prime_iff_prime_int at hp,
+  rw ← int.nat_cast_eq_coe_nat at *,
+  rw [pow_sub_pow' hp, pow_prime_pow_sub_pow_prime_pow hp hp1 hxy₁ hxy₂, enat.coe_get],
+  { rw ← geom_sum₂_mul,
+    exact dvd_mul_of_dvd_right hxy₁ _ },
+  { rw ← mul_pow,
+    intro h,
+    exact hxy₂ (hp.dvd_of_dvd_pow h) },
+  { iterate 2 { rw int.nat_cast_eq_coe_nat },
+    rw int.coe_nat_dvd,
```

Comment:
`simp only` is a good alternative to `rw` if you need to rewrite multiple times:
```suggestion
  { simp only [int.nat_cast_eq_coe_nat, int.coe_nat_dvd],
```

Reply:
I think I'd rather keep it the way it is